File: server/src/main/java/dev/ultreon/quantum/network/system/IConnection.java
package dev.ultreon.quantum.network.system;

import dev.ultreon.quantum.Logger;
import dev.ultreon.quantum.LoggerFactory;
import dev.ultreon.quantum.network.PacketHandler;
import dev.ultreon.quantum.network.PacketListener;
import dev.ultreon.quantum.network.packets.Packet;
import dev.ultreon.quantum.network.stage.PacketStage;
import dev.ultreon.quantum.network.stage.PacketStages;
import dev.ultreon.quantum.server.CloseCodes;
import dev.ultreon.quantum.server.player.ServerPlayer;
import dev.ultreon.quantum.text.TextObject;
import dev.ultreon.quantum.util.Result;

import javax.annotation.Nullable;
import java.io.Closeable;
import java.util.concurrent.atomic.AtomicInteger;

public interface IConnection<OurHandler extends PacketHandler, TheirHandler extends PacketHandler> extends Closeable {
    Logger LOGGER = LoggerFactory.getLogger("NetConnections");
    AtomicInteger rx = new AtomicInteger();
    AtomicInteger tx = new AtomicInteger();

    default void send(Packet<? extends TheirHandler> packet) {
        send(packet, null);
    }

    void send(Packet<? extends TheirHandler> packet, @Nullable PacketListener resultListener);

    boolean isCompressed();

    void disconnect(int code, String message);

    Result<Void> on3rdPartyDisconnect(int statusCode, String message);

    void queue(Runnable handler);

    void start();

    void moveTo(PacketStage stage, OurHandler handler);

    default boolean isConnecting() {
        return false;
    }

    boolean isConnected();

    default void tick() {

    }

    boolean isMemoryConnection();

    default void initiate(OurHandler handler, @Nullable Packet<? extends TheirHandler> packetToThem) {
        this.moveTo(PacketStages.LOGIN, handler);
        if (packetToThem != null)
            this.send(packetToThem);
    }

    void setReadOnly();

    void setPlayer(ServerPlayer player);

    default void disconnect(TextObject message) {
        disconnect(CloseCodes.NORMAL_CLOSURE.getCode(), message.getText());
    }

    long getPing();

    void update();

    void onPing(long ping);

    boolean isLoggingIn();
}

File: server/src/main/java/dev/ultreon/quantum/network/system/IConnection.kt
package dev.ultreon.quantum.network.system

import com.google.errorprone.annotations.CanIgnoreReturnValue
import dev.ultreon.quantum.log.Logger
import dev.ultreon.quantum.log.LoggerFactory
import dev.ultreon.quantum.network.PacketHandler
import dev.ultreon.quantum.network.PacketListener
import dev.ultreon.quantum.network.packets.Packet
import dev.ultreon.quantum.network.stage.PacketStage
import dev.ultreon.quantum.network.stage.PacketStages
import dev.ultreon.quantum.server.player.ServerPlayer
import dev.ultreon.quantum.text.TextObject
import dev.ultreon.quantum.util.Result
import java.io.Closeable
import java.util.concurrent.atomic.AtomicInteger

interface IConnection<OurHandler : PacketHandler, TheirHandler : PacketHandler> :
  Closeable {
  @CanIgnoreReturnValue
  fun send(packet: Packet<out TheirHandler>) {
    send(packet, null)
  }

  @Deprecated("")
  @CanIgnoreReturnValue
  fun send(packet: Packet<out TheirHandler>, resultListener: PacketListener?)

  val isCompressed: Boolean

  fun disconnect(message: String)

  fun on3rdPartyDisconnect(message: String): Result<Void>

  fun queue(handler: Runnable)

  fun start()

  fun moveTo(stage: PacketStage, handler: OurHandler)

  val isConnecting: Boolean
    get() = false

  val isConnected: Boolean

  fun tick() {
  }

  val isMemoryConnection: Boolean

  fun initiate(handler: OurHandler, packetToThem: Packet<out TheirHandler>?) {
    this.moveTo(PacketStages.LOGIN, handler)
    if (packetToThem != null) this.send(packetToThem)
  }

  fun setReadOnly()

  fun setPlayer(player: ServerPlayer?)

  fun disconnect(message: TextObject) {
    disconnect(message.text)
  }

  val ping: Long

  fun onPing(ping: Long)

  companion object {
    @JvmStatic
    val logger: Logger = LoggerFactory.getLogger("NetConnections")
    @JvmStatic
    val rx: AtomicInteger = AtomicInteger()
    @JvmStatic
    val tx: AtomicInteger = AtomicInteger()
  }
}
